### ISSUE TITLE
Alternative for Window Focus Switching

### ISSUE THREAD

**Issue by varsop** [material-shell/material-shell#233]
Whenever I have more than 3 windows tiled in a workspace I find switching focus of a window a bit irritating. Switching the focus of a window in MS works by using the list of windows displayed in the upper panel. 

Assume I have focused a window which has position number n in the panel list and I want to switch to a window which has position number n' and |n - n'| > 1. Furthermore assume window with number n and window with number n' share a border in a certain tiling layout and are therefore "neighbored" as windows in this tiling layout. Then it would be nice to jump over this border by just one key combination, instead of navigating |n - n'-times in the upper panel list. This is the window movement as pop shell has implemented. 

Thus, it would be nice to navigate by neighbored windows in the upper panel list and to navigate by neighbored windows in a certain tiling layout.

**Comment by PapyElGringo** [comment]
What is the issue with `super+a` and `super+d`?

**Comment by varsop** [comment]
> Ah you would like navigate by direction? but you will need 4 different hotkey shortcut for that ? left / right / top / bottom?

Exactly. Indeed, I would need 4 directions to jump over neighboring borders of windows. I know this kind of behavior from emacs, whenever I split a buffer and navigate via [wind move](https://www.emacswiki.org/emacs/WindMove)

**Comment by PapyElGringo** [comment]
It's currently too complicated since we don't have this kind of informations but will be doable when we will rework to implemant the resize feature since we should follow the logic of splitting in direction.

**Comment by varsop** [comment]
Okay, I understand. I put here the ideas of my desired behavior. I would consequently reorder the shortcuts for navigation, like this:

**Navigation for workspaces (left panel):**
Super + Page up: move one workspace up
Super + Page down: move one workspace down

**Navigation within upper panel:**  
Super + , :  move one window left      _"," is the key with the sign "<" on ANSI layout keyboard_
Super + . : move one window right     _"." is the key with the sign ">" on ANSI layout keyboard_

**Navigation via directions:**
We use the arrow keys indicating the direction of a movement. We also want to navigate by directions if all windows are maximized and the split layout is chosen.

 **All windows maximized**
Super + left arrow = Super + ,
Super + right arrow = Super + .
Super + up arrow: _does nothing_ 
Super + down arrow: _does nothing_ 

**Split layout**
Super + left arrow = Super + ,
Super + right arrow = Super + .
Super + up arrow: _does nothing_ 
Super + down arrow: _does nothing_ 

**Tiling layout, which is not one of the two above**
Super + left arrow:

_Switch focus to the window, which is the nearest left neighbor. Therefor we need to define the set of all left neighboring windows. A window is neighbored to another window, if they share an edge. We identify the left screen edge with the right screen edge. In this way windows with an edge touching the left screen side can be neighbored to windows with an edge touching the right side of the screen. Furthermore we need to define a distance between the cursor position in the current window to all the left neighboring windows. Choose the smallest on, if it exists. If there is more than one, choose a unique one by some special property._

Super + right arrow:

_Switch focus to the window, which is the nearest left neighbor. The explanation of the terminology can be taken from above, where the same definitions apply._

Super + up arrow:

_Switch focus to the window, which is the nearest upper neighbor. We identify the upper screen edge with the lower screen edge. In this way windows with an edge touching upper screen side can be neighbored to windows with an edge touching the lower screen side. The term nearest stems from the existence of a metric function between windows on a workspace_

Super + down arrow:

_Switch focus to the window, which is the nearest lower neighbor. Terminology like above_
